[SPARK-42162] Introduce MultiCommutativeOp expression as a memory optimization for canonicalizing large trees of commutative expressions

db-scnakandala opened the pull request:
### What changes were proposed in this pull request?
- This PR introduces a new expression called `MultiCommutativeOp` which is used by the commutative expressions (e.g., `Add`, `Multiply`, `And`, `Or`, `BitwiseOr`, `BitwiseAnd`, `BitwiseXor`) during canonicalization.
- During canonicalization, when there is a list of consecutive commutative expressions, we now create a MultiCommutative expression with references to original operands, instead of creating new objects.
- This new expression is added as a memory optimization to reduce generating a large number of intermediate objects during canonicalization.

### Why are the changes needed?
- With the [recent changes](https://github.com/apache/spark/pull/37851) in the expression canonicalization, a complex query with a large number of commutative operations could end up consuming significantly more (sometimes > 10X) memory on the executors.
- In our case, this issue happens for a specific complex query that has a huge expression tree containing Add operators interleaved by non Add operators.
- The issue is related to canonicalization and why it is causing issues in the executors is because the codegen component relies on expression canonicalization to deduplicate expressions. 
- When we have a large number of Adds interleaved by non-Add operators, [this line](https://github.com/apache/spark/pull/37851/files#diff-7278f2db37934522ee7c74b71525153234cff245cefaf996957e4a9ff3dbaacdR1171) ends up materializing a new canonicalized expression tree at every non-Add operator.
- In our case, analyzing the executor heap histogram shows that the additional memory is consumed by a large number of Add objects.
- The high memory usage causes the executors to lose heartbeat signals and results in task failures.
- The proposed `MultiCommutativeOp` expression avoids generating new Add expressions and keeps the extra memory usage to a minimum.


### Does this PR introduce _any_ user-facing change?
- No

### How was this patch tested?
- Existing unit tests and new unit tests.

db-scnakandala committed:
adding MultiAdd

peter-toth commented:
> - With the https://github.com/apache/spark/pull/37851 in the expression canonicalization, a complex query with a large number of Add operations could end up consuming significantly more (sometimes > 10X) memory on the executors.

@db-scnakandala, can you please explain this issue a bit more? Does https://github.com/apache/spark/pull/37851 cause performance regression? Why exactly? And why on executors?

db-scnakandala commented:
> > * With the [[SPARK-40362][SQL] Fix BinaryComparison canonicalization #37851](https://github.com/apache/spark/pull/37851) in the expression canonicalization, a complex query with a large number of Add operations could end up consuming significantly more (sometimes > 10X) memory on the executors.
> 
> @db-scnakandala, can you please explain this issue a bit more? Does #37851 cause performance regression? Why exactly? And why on executors?

@peter-toth This issue happens for a specific complex query that has a huge expression tree containing Add operators interleaved by non Add operators.

For this query, the query ends up consuming 10x more memory and ultimately results in executor failures.

The issue is related to canonicalization and why it is causing issues in the executors is because the codegen component relies on expression canonicalization to deduplicate expressions. 

When we have a large number of Adds interleaved by non-Add operatos, I think [this line](https://github.com/apache/spark/pull/37851/files#diff-7278f2db37934522ee7c74b71525153234cff245cefaf996957e4a9ff3dbaacdR1171) ends up materializing a new canonicalized expression tree at every non-Add operator.

In our case, analyzing the executor heap histogram shows that the additional memory is consumed by a large number of Add objects.

Sorry for not being clear in the initial PR description. I will update it now.

db-scnakandala committed:
conf disable test

db-scnakandala commented:
@cloud-fan @peter-toth This PR is ready for review now. Thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -407,6 +408,74 @@ object BinaryArithmetic {
+  // Result Precision: max(s1, s2) + max(p1-s1, p2-s2) + 1
+  // Result Scale:     max(s1, s2)
+  // scalastyle:on
+  private def resultDecimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType = {
```
shall we move it to a util function to avoid duplicating the code?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -240,6 +240,15 @@ object SQLConf {
     .createWithDefault(100)
 
+  val MULTI_ADD_OPT_THRESHOLD =
+    buildConf("spark.sql.analyzer.canonicalization.multiAddMemoryOptThreshold")
```
shall we apply it to similar operations like `Multiply`? seems we can make changes in `trait CommutativeExpression`

peter-toth commented:
> @peter-toth This issue happens for a specific complex query that has a huge expression tree containing Add operators interleaved by non Add operators.
> 
> For this query, the query ends up consuming 10x more memory and ultimately results in executor failures.
> 
> The issue is related to canonicalization and why it is causing issues in the executors is because the codegen component relies on expression canonicalization to deduplicate expressions.
> 
> When we have a large number of Adds interleaved by non-Add operatos, I think [this line](https://github.com/apache/spark/pull/37851/files#diff-7278f2db37934522ee7c74b71525153234cff245cefaf996957e4a9ff3dbaacdR1171) ends up materializing a new canonicalized expression tree at every non-Add operator.
> 
> In our case, analyzing the executor heap histogram shows that the additional memory is consumed by a large number of Add objects.
> 
> Sorry for not being clear in the initial PR description. I will update it now.

So this is very likely because subexpression elimination uses `EquivalentExpressions.addExprTree()` that recurses the expression tree and canonicalizes each and every node. Since canonicalization of a `CommutativeExpression` node doesn't reuse the canonicalized form of its `CommutativeExpression` children, each `Add` node in that nested group of `Add`s has a huge separate canonicalized tree.
The new `MultiAdd` looks like a good way to fix this issue, but as @cloud-fan mentioned, we should make it more general to handle other `CommutativeExpression`s too.

db-scnakandala committed:
addressing review comments

db-scnakandala commented:
@cloud-fan @peter-toth Thank you for the feedback. In the latest changes, I have made the approach more general. The optimization now gets applied to `Add`, `Multiply`, `And`, `Or`, `BitwiseAnd`, `BitwiseOr`, `BitwiseXor`.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -1335,3 +1335,70 @@ trait CommutativeExpression extends Expression {
+ *                     Multiply operations when the input datatype is decimal.
+ */
+case class MultiCommutativeOp(
+  operands: Seq[Expression],
```
nit: 4 spaces indentation

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -1335,3 +1335,70 @@ trait CommutativeExpression extends Expression {
+ */
+case class MultiCommutativeOp(
+  operands: Seq[Expression],
+  evalMode: Option[EvalMode.Value])(originalRoot: Expression) extends Unevaluable {
```
shall we add one more parameter `opCls: Class[_]`? Otherwise a bunch of `Add` can be equal to a bunch of `Multiply`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -1335,3 +1335,70 @@ trait CommutativeExpression extends Expression {
+   */
+  override def dataType: DataType = {
+    originalRoot match {
+      case Add(_, _, _) | Multiply(_, _, _) =>
```
```suggestion
      case _: Add | _: Multiply =>
```

db-scnakandala committed:
addressing review comments

db-scnakandala commented:
@cloud-fan @peter-toth Thank you for the feedback. I have addressed your comments in the latest version.
Please let me know if anything needs to be changed.

peter-toth commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -240,6 +240,15 @@ object SQLConf {
+      .internal()
+      .doc("The minimum number of consecutive commutative expressions to" +
+        " invoke the MultiCommutativeOp memory optimization during canonicalization.")
+      .version("3.3.0")
```
This version should be `3.5.0` or `3.4.0` depending on where we want to merge this PR.

peter-toth commented:
PR looks good to me, but I'm not sure why would anyone tune the new `spark.sql.analyzer.canonicalization.multiCommutativeOpMemoryOptThreshold` config.

db-scnakandala commented:
> PR looks good to me, but I'm not sure why would anyone tune the new

There are some downstream requirements where we want the canonicalized expressions of the analyzed to be evaluable for optimizations like constant folding. The MultiCommutativeOp is not evaluable. Hence, having the conf enables us to apply the optimization only for very large expressions.

Maybe we can change the default value to 2. WDYT?

peter-toth commented:
> Maybe we can change the default value to 2. WDYT?

Now the description of the config says the `The minimum number of consecutive commutative expressions`, but seemingly the logic uses the number of non-commutative operads below those commutative expressions (`operands.length < SQLConf.get.getConf(MULTI_COMMUTATIVE_OP_OPT_THRESHOLD)`). So I would argue that the default value of 2 looks good, but the logic should follow that description and we should optimize only when there are at least 2 nested expressions and shouldn't canonicalize a single expression to `MultiCommutativeOp`.

db-scnakandala committed:
changing the default value

db-scnakandala commented:
> > Maybe we can change the default value to 2. WDYT?
> 
> Now the description of the config says the `The minimum number of consecutive commutative expressions`, but seemingly the logic uses the number of non-commutative operads below those commutative expressions (`operands.length < SQLConf.get.getConf(MULTI_COMMUTATIVE_OP_OPT_THRESHOLD)`). So I would argue that the default value 2 sounds good, but the logic should follow that description and we should optimize only when there are at least 2 nested expressions and shouldn't canonicalize a single expression to `MultiCommutativeOp`.

@peter-toth Can you elaborate more on the statement "shouldn't canonicalize a single expression to `MultiCommutativeOp`"

From what I understand since the operator is a commutative expression, it should at least have two operands (?)

peter-toth commented:
> @peter-toth Can you elaborate more on the statement "shouldn't canonicalize a single expression to MultiCommutativeOp"
> 
> From what I understand since the operator is a commutative expression, it should at least have two operands (?)

I would keep the canonicalization of a single commutative expression like: `Add(x, y)` as it is: `Add(x, y)`, but when we have at least 2 nested: `Add(Add(x, y), z)` then as `MultiCommutativeOp(x, y, z, Add)`

db-scnakandala commented:
> > @peter-toth Can you elaborate more on the statement "shouldn't canonicalize a single expression to MultiCommutativeOp"
> > From what I understand since the operator is a commutative expression, it should at least have two operands (?)
> 
> I would keep the canonicalization of a single commutative expression like: `Add(x, y)` as it is: `Add(x, y)`, but when we have at least 2 nested: `Add(Add(x, y), z)` then as `MultiCommutativeOp(x, y, z, Add)`

If I understand it correctly, setting the threshold to `3` should achieve this requirement, right?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -1335,3 +1335,72 @@ trait CommutativeExpression extends Expression {
+
+  /**
+   * Returns the [[DataType]] of the result of evaluating this expression.  It is
+   * invalid to query the dataType of an unresolved expression (i.e., when `resolved` == false).
```
nit: if you look at other sub-classes of `Expression`, we don't repeat the api doc in the override methods.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -240,6 +240,15 @@ object SQLConf {
+  val MULTI_COMMUTATIVE_OP_OPT_THRESHOLD =
+    buildConf("spark.sql.analyzer.canonicalization.multiCommutativeOpMemoryOptThreshold")
+      .internal()
+      .doc("The minimum number of consecutive non-commutative operands to" +
```
non-commutative?

db-scnakandala replied:
Changed it to `The minimum number of operands in a commutative expression tree`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -1335,3 +1335,72 @@ trait CommutativeExpression extends Expression {
+   * Children should not change. Immutability required for containsChild optimization
+   */
+  override def children: Seq[Expression] = operands
+
```
how about providing a util function in this trait so that sub-classes is easier to do canonicalization?
```
def buildCanonicalizedPlan(collectOperands: PartialFunction[Expression, Seq[Expression]], buildBinaryOp: (Expression, Expression) => Expression) = ...
```

db-scnakandala committed:
adressing review comments

db-scnakandala committed:
refactor

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -632,7 +636,12 @@ case class Multiply(
   override lazy val canonicalized: Expression = {
     // TODO: do not reorder consecutive `Multiply`s with different `evalMode`
-    orderCommutative({ case Multiply(l, r, _) => Seq(l, r) }).reduce(Multiply(_, _, evalMode))
+    val reorderResult = buildCanonicalizedPlan(
+      { case Multiply(l, r, _) => Seq(l, r) }
+      { case (l: Expression, r: Expression) => Multiply(l, r, evalMode)},
+      Some(evalMode)
+    )
+    reorderResult
```
```suggestion
    buildCanonicalizedPlan(
      { case Multiply(l, r, _) => Seq(l, r) }
      { case (l: Expression, r: Expression) => Multiply(l, r, evalMode)},
      Some(evalMode)
    )
```

cloud-fan replied:
same for other places.

db-scnakandala replied:
done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -479,8 +480,11 @@ case class Add(
+      { case (l: Expression, r: Expression) => Add(l, r, evalMode)},
+      Some(evalMode)
+    )
     if (resolved && reorderResult.resolved && reorderResult.dataType == dataType) {
```
not related to this PR, @gengliangwang do we have a followup to apply this fix for multiply?

db-scnakandala committed:
removing redudant val

peter-toth commented:
> > I would keep the canonicalization of a single commutative expression like: `Add(x, y)` as it is: `Add(x, y)`, but when we have at least 2 nested: `Add(Add(x, y), z)` then as `MultiCommutativeOp(x, y, z, Add)`
> 
> If I understand it correctly, setting the threshold to `3` should achieve this requirement, right?

You changed the config semantics from the number of commutative expressions to the number of operands in commutative expression tree (https://github.com/apache/spark/pull/39722/commits/e196e7d60d664af2502faab212477d24c2398d49), so now 3 is the value I too prefer.
This is because the memory footprint of `Add(x, y)` seems smaller than `MultiCommutativeOp(x, y, Add)`, but this changes when have at least 2 `Add`s.

cloud-fan commented:
thanks, merging to master/3.4 (it's a regression)!

peter-toth commented:
Actually, I'm not sure this is a regression. The canonicalization process was different before https://github.com/apache/spark/pull/37851 but the canonicalized form of the nodes were the same. As far as I see it had been even more memory consuming as there was a `lazy val preCanonicalized` field in each node.
But I think it is nice to have this fix in 3.3 too.
